web/src/app/admin/connector/[ccPairId]/IndexingAttemptsTable.tsx

```diff
@@ -8,31 +10,149 @@ import {
   TableBody,
   TableCell,
   Text,
-  Button,
-  Divider,
 } from "@tremor/react";
-import { IndexAttemptStatus } from "@/components/Status";
 import { CCPairFullInfo } from "./types";
+import { IndexAttemptStatus } from "@/components/Status";
 import { useState } from "react";
 import { PageSelector } from "@/components/PageSelector";
+import { ThreeDotsLoader } from "@/components/Loading";
+import { buildCCPairInfoUrl } from "./lib";
 import { localizeAndPrettify } from "@/lib/time";
 import { getDocsProcessedPerMinute } from "@/lib/indexAttempt";
-import { Modal } from "@/components/Modal";
-import { CheckmarkIcon, CopyIcon, SearchIcon } from "@/components/icons/icons";
+import { errorHandlingFetcher } from "@/lib/fetcher";
+import { ErrorCallout } from "@/components/ErrorCallout";
+import { SearchIcon } from "@/components/icons/icons";
 import Link from "next/link";
 import ExceptionTraceModal from "@/components/modals/ExceptionTraceModal";
+import { PaginatedIndexAttempts } from "./types";
+import { useRouter } from "next/navigation";
 
 const NUM_IN_PAGE = 8;
+const PAGES_TO_PREFETCH = 20;
 
 export function IndexingAttemptsTable({ ccPair }: { ccPair: CCPairFullInfo }) {
-  const [page, setPage] = useState(1);
+  const router = useRouter();
+  const [page, setPage] = useState(() => {
+    if (typeof window !== "undefined") {
+      const urlParams = new URLSearchParams(window.location.search);
+      return parseInt(urlParams.get("page") || "1", 10);
+    }
+    return 1;
+  });
   const [indexAttemptTracePopupId, setIndexAttemptTracePopupId] = useState<
     number | null
   >(null);
-  const indexAttemptToDisplayTraceFor = ccPair.index_attempts.find(
+  const [currentPageData, setCurrentPageData] =
```

Comment:
I cooked from here down. 🧑‍🍳 
The main features are:

- preloading pages
- debouncing (in the form of a 200 ms delay from clicking the button to actually calling the api)
- instant loading of preloaded pages

Reply:
READ THE COMMENTS FOR EACH PART, THEY ARE INFORMATIVE